tests/shell_cmds_tests.cpp

```diff
@@ -171,25 +173,114 @@ TEST_F(DOS_Shell_CMDSTest, CMD_ECHO_space_handling)
 	EXPECT_FALSE(shell.echo);
 
 	// NOTE: the expected string here is missing the leading char of the
 	// input to ECHO. the first char is stripped as it's assumed it will be
 	// a space, period or slash.
 	EXPECT_CALL(shell, WriteOut(_, StrEq("    HI "))).Times(1);
-	EXPECT_NO_THROW({ shell.CMD_ECHO(const_cast<char *>(".    HI ")); });
+	EXPECT_NO_THROW({ shell.CMD_ECHO(const_cast<char*>(".    HI ")); });
+}
+
+TEST_F(DOS_Shell_CMDSTest, CMD_FOR_basic)
+{
+	MockDOS_Shell shell = {};
+
+	EXPECT_CALL(shell, ExecuteShellCommand(StrEq("ECHO"), StrEq(" ONE")))
+	        .Times(1)
+	        .WillOnce(Return(true));
+	EXPECT_CALL(shell, ExecuteShellCommand(StrEq("ECHO"), StrEq(" TWO")))
+	        .Times(1)
+	        .WillOnce(Return(true));
+	EXPECT_NO_THROW({
+		shell.CMD_FOR(const_cast<char*>(" %C IN (ONE TWO) DO ECHO %C"));
+	});
+}
+
+TEST_F(DOS_Shell_CMDSTest, CMD_FOR_delimiters)
+{
+	MockDOS_Shell shell = {};
+
+	static constexpr auto delimeters = ",;= \t";
+	// TODO: C++20: Use std::format();
+	const auto input = format_string("%s%%C%sIN%s(%sONE%sTWO%s)%sDO%sECHO %%C",
+	                                 delimeters,
+	                                 delimeters,
+	                                 delimeters,
+	                                 delimeters,
+	                                 delimeters,
+	                                 delimeters,
+	                                 delimeters,
+	                                 delimeters); // ):
+
+	EXPECT_CALL(shell, ExecuteShellCommand(StrEq("ECHO"), StrEq(" ONE")))
+	        .Times(1)
+	        .WillOnce(Return(true));
+	EXPECT_CALL(shell, ExecuteShellCommand(StrEq("ECHO"), StrEq(" TWO")))
+	        .Times(1)
+	        .WillOnce(Return(true));
+	EXPECT_NO_THROW({ shell.CMD_FOR(const_cast<char*>(input.c_str())); });
+}
+
+TEST_F(DOS_Shell_CMDSTest, CMD_FOR_missing_do)
+{
+	MockDOS_Shell shell = {};
+
+	EXPECT_CALL(shell, ExecuteShellCommand(_, _)).Times(0);
+	EXPECT_NO_THROW({
+		shell.CMD_FOR(const_cast<char*>(" %C IN (ONE TWO) ECHO %C"));
+	});
+}
+
+TEST_F(DOS_Shell_CMDSTest, CMD_FOR_missing_in)
+{
+	MockDOS_Shell shell = {};
+
+	EXPECT_CALL(shell, ExecuteShellCommand(_, _)).Times(0);
+	EXPECT_NO_THROW({
+		shell.CMD_FOR(const_cast<char*>(" %C (ONE TWO) DO ECHO %C"));
+	});
+}
+
+TEST_F(DOS_Shell_CMDSTest, CMD_FOR_missing_var)
+{
+	MockDOS_Shell shell = {};
+
+	EXPECT_CALL(shell, ExecuteShellCommand(_, _)).Times(0);
+	EXPECT_NO_THROW({
+		shell.CMD_FOR(const_cast<char*>(" IN (ONE TWO) DO ECHO %C"));
+	});
+}
+
+TEST_F(DOS_Shell_CMDSTest, CMD_FOR_missing_parens)
+{
+	MockDOS_Shell shell = {};
+
+	EXPECT_CALL(shell, ExecuteShellCommand(_, _)).Times(0);
+	EXPECT_NO_THROW({
+		shell.CMD_FOR(const_cast<char*>(" %C IN ONE TWO DO ECHO %C"));
+	});
+}
+
+TEST_F(DOS_Shell_CMDSTest, CMD_FOR_missing_command)
+{
+	MockDOS_Shell shell = {};
+
+	EXPECT_CALL(shell, ExecuteShellCommand(_, _)).Times(0);
+	EXPECT_NO_THROW(
+	        { shell.CMD_FOR(const_cast<char*>(" %C IN (ONE TWO) DO")); });
```

Comment:
Was playing around with random FOR loops in `4DOS`, and packaged them up: [for_test_cases.zip](https://github.com/dosbox-staging/dosbox-staging/files/13321772/for_test_cases.zip)

```
for_test_cases/
├── 4dos.com
├── bad
│   ├── for_corrupt_1.bat
│   ├── for_corrupt_2.bat
│   ├── for_corrupt_3.bat
│   ├── for_corrupt_4.bat
│   ├── for_missing_end_paren.bat
│   └── for_missing_start_paren.bat
└── good
    ├── for_empty_body.bat
    ├── for_empty_vars.bat
    ├── for_files_multiline.bat
    ├── for_implied_var.bat
    ├── for_nested_loops.bat
    └── for_vars_multiline.bat
```


Maybe these could dropped into the unit tests? (I guess as raw multi-line string literals?)

Reply:
That would be very nice, indeed.

Reply:
Ah, I missed that. We want to stick to MS-DOS 6.22 compatibility, yeah.

Reply:
100% agree; sorry - I should have been more careful and tracked down a DOS 6.22 COMMAND.COM to test these with.